server/routes/api/collections/collections.ts

```diff
@@ -881,6 +892,187 @@ router.post(
   }
 );
 
+router.post(
+  "collections.archive",
+  auth(),
+  validate(T.CollectionsArchiveSchema),
+  transaction(),
+  async (ctx: APIContext<T.CollectionsArchiveReq>) => {
+    const { transaction } = ctx.state;
+    const { id } = ctx.input.body;
+    const { user } = ctx.state.auth;
+
+    const collection = await Collection.scope([
+      {
+        method: ["withMembership", user.id],
+      },
+    ]).findByPk(id, {
+      transaction,
+    });
+
+    if (!collection) {
+      throw NotFoundError("Collection not found");
+    }
+
+    authorize(user, "archive", collection);
+
+    collection.archivedAt = new Date();
+    collection.archivedById = user.id;
+    await collection.save({ transaction });
+    collection.archivedBy = user;
+
+    // Archive all documents within the collection
+    await Document.update(
+      {
+        lastModifiedById: user.id,
+        archivedAt: collection.archivedAt,
+      },
+      {
+        where: {
+          teamId: collection.teamId,
+          collectionId: collection.id,
+          archivedAt: {
+            [Op.is]: null,
+          },
+        },
+        transaction,
+      }
+    );
+
+    await Event.createFromContext(
+      ctx,
+      {
+        name: "collections.archive",
+        collectionId: collection.id,
+        data: {
+          name: collection.name,
+          archivedAt: collection.archivedAt,
+        },
+      },
+      { transaction }
+    );
+
+    ctx.body = {
+      data: await presentCollection(ctx, collection),
+      policies: presentPolicies(user, [collection]),
+    };
+  }
+);
+
+router.post(
+  "collections.restore",
+  auth(),
+  validate(T.CollectionsRestoreSchema),
+  transaction(),
+  async (ctx: APIContext<T.CollectionsRestoreReq>) => {
+    const { transaction } = ctx.state;
+    const { id } = ctx.input.body;
+    const { user } = ctx.state.auth;
+
+    const collection = await Collection.scope({
+      method: ["withMembership", user.id],
+    }).findByPk(id, {
+      transaction,
+    });
+
+    if (!collection) {
+      throw NotFoundError("Collection not found");
+    }
+
+    authorize(user, "restore", collection);
+
+    const collectionArchivedAt = collection.archivedAt;
+
+    await Document.update(
+      {
+        lastModifiedById: user.id,
+        archivedAt: null,
+      },
+      {
+        where: {
+          collectionId: collection.id,
+          teamId: user.teamId,
+          archivedAt: collection.archivedAt,
```

Comment:
Wanted to especially highlight this! The condition here is put with the intention that _only_ the docs which were archived as part of collection archival will be restored. Although there's a catch in a _rare_ scenario where a doc within a collection was _separately_ archived at the _same time_ the collection was archived – in this case, both might end up having the _same_ `archivedAt`. Then, if the collection is restored, the doc which was separately archived, would be restored as part of collection restoration, which would be an unintended behavior.

Reply:
Maybe _locking_ rows and then setting `archivedAt` to `transaction_timestamp()` could be a solution.

Reply:
>  there's a catch in a rare scenario where a doc within a collection was separately archived at the same time the collection was archived

I suppose I don't understand this, how would the timestamps be identical?

Reply:
Suppose a user triggered `documents.archive` for a document with id = `docId` and _concurrently_ `collections.archive` for the collection to which `docId` belongs. In a rare scenario, this might end up setting `archivedAt`, for both the doc and collection, to same timestamp. Then 👇 
>if the collection is restored, the doc which was separately archived, would be restored as part of collection restoration, which would be an unintended behavior.

Reply:
But it would have to be the identical millisecond? Or am I missing something there

Reply:
Yeah, exact same `new Date()` value for both. Hence, extremely rare.